apps/test/unit/lib/levelbuilder/lesson-editor/UploadImageDialogTest.jsx

```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import sinon from 'sinon';
+import {shallow} from 'enzyme';
+
+import UploadImageDialog from '@cdo/apps/lib/levelbuilder/lesson-editor/UploadImageDialog';
+
+import {expect} from '../../../../util/reconfiguredChai';
+
+describe('UploadImageDialog', () => {
+  it('uploads selected image', () => {
+    const wrapper = shallow(
+      <UploadImageDialog
+        isOpen={true}
+        handleClose={() => {}}
+        uploadImage={() => {}}
+      />
+    );
+
+    const fetchStub = sinon.stub(window, 'fetch').resolves();
+    wrapper.instance().handleChange({target: {files: ['filedata']}});
+
+    expect(fetchStub.callCount).to.equal(1);
+
+    const fetchCall = fetchStub.getCall(0);
+    expect(fetchCall.args[0]).to.equal('/level_assets/upload');
+    expect(fetchCall.args[1].body.get('file')).to.equal('filedata');
+
+    fetchStub.restore();
+  });
+
+  it('returns the uploaded image url', () => {
+    const handleClose = sinon.fake();
+    const uploadImage = sinon.fake();
+    const wrapper = shallow(
+      <UploadImageDialog
+        isOpen={true}
+        handleClose={handleClose}
+        uploadImage={uploadImage}
+      />
+    );
+    wrapper.setState({imgUrl: 'http://example.com/img.png'});
+
+    expect(handleClose.callCount).to.equal(0);
+    expect(uploadImage.callCount).to.equal(0);
+    wrapper.instance().handleCloseAndSave();
+
+    expect(handleClose.callCount).to.equal(1);
+    expect(uploadImage.callCount).to.equal(1);
+    expect(uploadImage.calledWith('http://example.com/img.png')).to.equal(true);
+  });
+});
```

Comment:
this looks like good test coverage. I am assuming that you are not using sinon.fakeServer here because it does not stub the fetch API. what do you think about (not right now but maybe someday) using a library like [fake-fetch](https://www.npmjs.com/package/fake-fetch) that lets us stub the fetch API more like how we use sinon.fakeServer to stub XHRs?

Reply:
ooh, actually I didn't use it because I didn't know about it.

I was actually going to specifically call out my stubbing of fetch in this test as an example of the kind of testing of implementation details that I was just telling you yesterday we should avoid. The test shouldn't care if we're specifically using fetch or not, just that we're making some kind of post; `fakeServer` looks like it might be perfect for this.

I'm not familiar with `fake-fetch`; how is it different from what I'm doing here?

Reply:
my understanding on this is a bit theoretical, but here it goes. Using this terminology: https://en.wikipedia.org/wiki/Test_double, the difference is that what you've got here is a "mock". sinon.fakeServer is a "fake", meaning you tell it "the server responds to X with Y", rather than "the method will be called with parameters XYZ". my understanding is that fake-fetch is also a fake.

Here is an example of fakeServer being used: https://github.com/code-dot-org/code-dot-org/blob/ad47df75fea734f3f8b638808c52dd49b124098a/apps/test/unit/lib/levelbuilder/lesson-editor/LessonEditorTest.js#L279-L283

The potential benefits of using a fake are:
1. make it easier to test end-to-end. using a fake in this PR, you could say "the server responds to X with Y", and stitch your two test cases together into one. I'm not sure if you want this, or what it would look like exactly with fake-fetch, or if you could perhaps already do this using our current approach. but my understanding was that fakeServer and fake-fetch would make this easier.
2. less boilerplate code. again I'd have to see fake-fetch actually being used to know that it reduces the number of lines or improves readability, but based on how fakeServer cleans things up, I am optimistic.

I hope this helps! Mostly I guess I have seen fakeServer clean up test code which covers XHRs, so was thinking that fake-fetch might do the same for the fetch API.